src/ng/parse.js

```diff
@@ -622,15 +622,43 @@ function isStateless($filter, filterName) {
   return !fn.$stateful;
 }
 
-function findConstantAndWatchExpressions(ast, $filter) {
+// Detect nodes which could depend on non-shallow state of objects
+function isPure(node, parentPureness) {
+  switch (node.type) {
+    // Computed members might invoke a stateful toString()
+    case AST.MemberExpression:
+      if (node.computed) {
+        return false;
+      }
+      break;
+
+    case AST.UnaryExpression:
+      return true;
```

Comment:
The idea is to allow shallow-watching the `input` in things like `fn(!input)` where `!parentIsPure`, but the child is.

I've tried to exclude case which can invoke `toString` like `"" + input` because those could be non-pure (when `toString` reads inner state).